dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/preheat/supplier/UniqueAttributeSupplierTest.java

```diff
@@ -0,0 +1,274 @@
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.trackedentity.TrackedEntityAttributeService;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentityattributevalue.TrackedEntityAttributeValue;
+import org.hisp.dhis.trackedentityattributevalue.TrackedEntityAttributeValueService;
+import org.hisp.dhis.tracker.TrackerImportParams;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+
+/**
+ * @author Enrico Colasante
+ */
+@ExtendWith( MockitoExtension.class )
+class UniqueAttributeSupplierTest extends DhisConvenienceTest
+{
+
+    private static final String UNIQUE_VALUE = "unique value";
+
+    private static final String TEI_UID = "TEI UID";
+
+    private static final String ANOTHER_TEI_UID = "ANOTHER_TEI_UID";
+
+    @InjectMocks
+    private UniqueAttributesSupplier supplier;
+
+    @Mock
+    private TrackedEntityAttributeService trackedEntityAttributeService;
+
+    @Mock
+    private TrackedEntityAttributeValueService trackedEntityAttributeValueService;
+
+    private TrackerImportParams params;
+
+    private TrackedEntityAttribute uniqueAttribute;
+
+    private TrackedEntityInstance tei;
+
+    private ProgramInstance programInstance;
+
+    private TrackedEntityAttributeValue trackedEntityAttributeValue;
+
+    @BeforeEach
+    public void setUp()
+    {
+        params = TrackerImportParams.builder().build();
+        uniqueAttribute = createTrackedEntityAttribute( 'A', ValueType.TEXT );
+        OrganisationUnit orgUnit = createOrganisationUnit( 'A' );
+        Program program = createProgram( 'A' );
+        Attribute attribute = createAttribute( 'A' );
+        AttributeValue attributeValue = createAttributeValue( attribute, UNIQUE_VALUE );
+        tei = createTrackedEntityInstance( 'A', orgUnit );
+        tei.setUid( TEI_UID );
+        tei.setAttributeValues( Collections.singleton( attributeValue ) );
+        programInstance = createProgramInstance( program, tei, orgUnit );
+        programInstance.setAttributeValues( Collections.singleton( attributeValue ) );
+        trackedEntityAttributeValue = createTrackedEntityAttributeValue( 'A', tei, uniqueAttribute );
+    }
+
+    @Test
+    void verifySupplierWhenNoUniqueAttributeIsPresentInTheSystem()
+    {
+        // given
+        TrackerPreheat preheat = new TrackerPreheat();
+        when( trackedEntityAttributeService.getAllUniqueTrackedEntityAttributes() )
+            .thenReturn( Collections.emptyList() );
+
+        // when
```

Comment:
This might be a matter of style, I would not add the comments for given, when, then.
For me using newlines and grouping the code into arrange, act, assert is sufficient to see what is being tested and what the expectations are.

Reply:
Done